# What drives the price of a car?

![](images/kurt.jpeg)

**OVERVIEW**

In this application, you will explore a dataset from kaggle. The original dataset contained information on 3 million used cars. The provided dataset contains information on 426K cars to ensure speed of processing.  Your goal is to understand what factors make a car more or less expensive.  As a result of your analysis, you should provide clear recommendations to your client -- a used car dealership -- as to what consumers value in a used car.

## CRISP-DM Framework

<center>
    <img src = images/crisp.png width = 50%/>
</center>


To frame the task, throughout our practical applications we will refer back to a standard process in industry for data projects called CRISP-DM.  This process provides a framework for working through a data problem.  Your first step in this application will be to read through a brief overview of CRISP-DM [here](https://mo-pcco.s3.us-east-1.amazonaws.com/BH-PCMLAI/module_11/readings_starter.zip).  After reading the overview, answer the questions below.

In [106]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures, StandardScaler, OneHotEncoder
from sklearn.compose import make_column_transformer

import pandas as pd

## Business Understanding

From a business perspective, we are tasked with identifying key drivers for used car prices.  In the CRISP-DM overview, we are asked to convert this business framing to a data problem definition.  Using a few sentences, reframe the task as a data task with the appropriate technical vocabulary. 

### Business Objectives
#### Background
The client is a used car dealership, an organization involved with the purchasing and reselling of used cars from and to customers.

#### Business Objectives
Based on the available dataset, we want to determine what factors play the most significant role in determining the price that a customer is willing to pay for a given used car. In doing so, we hope to answer the question "What factors should we focus on in regards to car inventory acquisition in order to improve revenue from used car sells?"

#### Business Success Criteria
Taking the provided dataset as the current state of car inventory, business success criteria here can be described as "if we were to focus on increasing inventory based on factors deemed most important to price, we should expect an increase in revenue performance". That is, success of this analysis would be actionable insights that allow the client to make changes to their inventory strategy and ultimately lead to an improvement in used car sales.

### Situation Assessment
#### Inventory of Resources
Resources available for this analysis are quite limited. No personnel are available for feedback, and have only one iteration of static data available for analysis.

Further, only one local computer is available, significantly limiting the amount of computational resources available. 

#### Requirements, Assumptions, and Constraints
In regards to resources, one of the major constraints is the limit of time. For this project, we have slightly under one week of time available, with the actual man-hours availible as a small fraction of this time period due to other previously agreed to committments to other engagements. As such, we will need to be very conservative in our choice of features to examine and in the time committed to model training. With this in mind, we should view this initial cycle of analysis as an MVP meant for the purpose of initial strategy assessment, with further cycles focusing more in-depth on identified key factors and allocating more time and resources as necessary.

In regards to the dataset available, we make the assumption that the data provided is sufficiently recent enough to provide an accurate image of the current state of inventory and its performance. Further, while the data entry process is unkown, we assume that minimal faulty data has been entered such that the data can be confidently relied on for accurate analysis.

We assume that the client maintains the legal rights to the dataset provided and that we are free to use it for any internally facing (eg. used/viewed only by the client for strategy reassessment) purposes.

#### Risks and Contingencies
|                 **Risk**                |                                                                                                                                                              **Contingency**                                                                                                                                                             |
|:---------------------------------------:|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
| A significant amount of data is invalid | Given that we are working with a dataset of over 400K cars the risk of this is low, however if this proves to be the case we will need to make a note of the reliability of the produced model and follow up in future cycles with better data.                                                                                          |
| Model chosen is suboptimal              | Given current limitations on available time, resources, and knowledge of regression models, there is a significant risk that the model chosen for evaluation will not be the best possible model available. The best way to handle this is to make note of model performance while making suggestions for future avenues of improvement. |
| Incorrect understanding of dataset      | Given that no feedback is available during the analysis cycle, there is a risk that data in the dataset may be incorrectly interpreted. Contingency for this is to spend sufficient time developing an understanding of the dataset and to go over understanding of data with client during the review phase.                            |

#### Costs and Benefits
While the current revenue of the client is currently unclear at this stage, given that the dataset provided for analysis has over 400K cars and the original dataset had over 3 million cars, it is not unreasonable to assume a significant benefit from this initial analysis.

For example, [using the following article as a base of reference](https://www.sapling.com/12129768/much-money-average-used-car-dealership-make-year), if we were to assume annual sales of cars total about \\$3,900,000, a conservative number given the age of the data referenced in the article, and thus about \\$325,000/month, a 5% improvement in sales would result in revenue of \\$341,250/month, or an increase of \\$16,250/month. This more than justifies the cost of a rough initial analysis.

### Data Mining Goals
Deliverables for this project include a notebook with all aspects of the CRISP-DM framework such as data cleaning, model training, and evaluation. Additionally, a README will be included summarizing findings in an easy to read format for non-technical consumers of this analysis.

The data will be made available via a Github Repo.

#### Data Mining Success Criteria
Key factors relating to used cars have been identified such that next steps relating to inventory strategy can be made.

### Project Plan
The first stage of this project involves an assessment of the dataset available. Given that we have a fixed timeline and no opportunity for follow-up during the current cycle, we will have to make due with whatever data is made available, making note of any issues as a point of follow-up for future analysis cycles in the initial data reports.

The next step is preparing our data for modeling. This involves cleaning the data, removing and creating any features as necesssary, and a report summarizing what steps have been taken to clean the data for future replication.

We will then perform an initial pass on model training, which, after evaluation of the trained model, may result in further passes to improve performance. Iterations here will likely be limited given the time constrains of the current cycle.

After modeling and evaluation, a summary of the work performed will be created along with key insights and next step recommendations.

## Data Understanding

After considering the business understanding, we want to get familiar with our data.  Write down some steps that you would take to get to know the dataset and identify any quality issues within.  Take time to get to know the dataset and explore what information it contains and how this could be used to inform your business understanding.

### Initial Data Collection
- Dataset: "vehicles.csv"
- Location: data/vehicles.csv
- Acquisition: Provided by client

In [2]:
df = pd.read_csv('data/vehicles.csv')

### Data Description
Examine collected data at a high level to gain an initial understanding of what kind of dataset is being utilized.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model         421603 non-null  object 
 6   condition     252776 non-null  object 
 7   cylinders     249202 non-null  object 
 8   fuel          423867 non-null  object 
 9   odometer      422480 non-null  float64
 10  title_status  418638 non-null  object 
 11  transmission  424324 non-null  object 
 12  VIN           265838 non-null  object 
 13  drive         296313 non-null  object 
 14  size          120519 non-null  object 
 15  type          334022 non-null  object 
 16  paint_color   296677 non-null  object 
 17  state         426880 non-null  object 
dtypes: f

In [5]:
df.head()

id                  region  price  year manufacturer model  \
0  7222695916                prescott   6000   NaN          NaN   NaN   
1  7218891961            fayetteville  11900   NaN          NaN   NaN   
2  7221797935            florida keys  21000   NaN          NaN   NaN   
3  7222270760  worcester / central MA   1500   NaN          NaN   NaN   
4  7210384030              greensboro   4900   NaN          NaN   NaN   

  condition cylinders fuel  odometer title_status transmission  VIN drive  \
0       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
1       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
2       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
3       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
4       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   

  size type paint_color state  
0  NaN  NaN         NaN    az  
1  NaN  NaN         NaN    ar  
2  NaN  NaN         NaN    fl  
3  NaN  NaN         NaN    ma  
4  NaN  NaN         NaN    nc

In [4]:
df.describe()

id         price           year      odometer
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07

#### Inital Impressions
Initial analysis reveals that the majority of features included as part of data are categorical. That is, values are within a predefined set of values. The features will need further encoding before they can be used for model training purposes.

Beyond these categorical features are four features with numerical values, **"id", "price", "year", and "odometer"**. **"id"** can be safely removed as it duplicates the function of the default index value with no added benefit. **"price"** is the dependent value and will need to be removed from the training data. **"year** and **odometer** are likely to serve as good features for model training.

#### Unclear Features
Initial examination of features leads to the conclusion that the provided features and their data are clear enough to understand what purpose they serve. More in-depth examination of data included per feature is required to see if there are any unusual/unclear values.

#### Missing Values
A number of features include missing values, with **"size"** missing the most with only 120,519/426,880, or approximately only 28% of registered cars including data. How to handle this missing data will require further investigation in order to answer questions such as "why is this data missing", "what are common values for this missing data", and "can data be manipulated to fill in these missing values". 

While it is advantageous to preserve as many data samples as possible, reducing to the minimum fully registered dataset of 120,519 would still represent a significant number of samples to train a model from, and as such leaves the initial impression that there is minimal risk of insufficient significant data.

### Data Exploration
#### Exploration of Available Values
##### Region

In [23]:
df.region.value_counts()

columbus                   3608
jacksonville               3562
spokane / coeur d'alene    2988
eugene                     2985
fresno / madera            2983
                           ... 
meridian                     28
southwest MS                 14
kansas city                  11
fort smith, AR                9
west virginia (old)           8
Name: region, Length: 404, dtype: int64

While the region for a given car is likely to have an impact on the price, this is not a feature that the client would be able to have any control over. As such, for the purpose of reducing data complexity, this feature can be dropped.

##### Manufacturer

In [8]:
df.manufacturer.value_counts()

ford               70985
chevrolet          55064
toyota             34202
honda              21269
nissan             19067
jeep               19014
ram                18342
gmc                16785
bmw                14699
dodge              13707
mercedes-benz      11817
hyundai            10338
subaru              9495
volkswagen          9345
kia                 8457
lexus               8200
audi                7573
cadillac            6953
chrysler            6031
acura               5978
buick               5501
mazda               5427
infiniti            4802
lincoln             4220
volvo               3374
mitsubishi          3292
mini                2376
pontiac             2288
rover               2113
jaguar              1946
porsche             1384
mercury             1184
saturn              1090
alfa-romeo           897
tesla                868
fiat                 792
harley-davidson      153
ferrari               95
datsun                63
aston-martin          24


In [20]:
# Manufacturers that represent at least 2% of the total dataset
manufacturer_group = df.groupby('manufacturer').manufacturer.count()
manufacturer_group[manufacturer_group > (df.shape[0] * 0.02)].sort_values(ascending = False)

manufacturer
ford             70985
chevrolet        55064
toyota           34202
honda            21269
nissan           19067
jeep             19014
ram              18342
gmc              16785
bmw              14699
dodge            13707
mercedes-benz    11817
hyundai          10338
subaru            9495
volkswagen        9345
Name: manufacturer, dtype: int64

In [22]:
# Total cars represented by manufacturers exceeding %2 of total dataset
manufacturer_group.values.sum()

409234

Let's limit analysis to cars within these manufacturers in order to reduce noise from cars that are not significantly represented in the dataset. Further exploration should be limited to only cars that exceed the 2% representation threshold.

In [26]:
manufacturer_list = manufacturer_group[manufacturer_group > (df.shape[0] * 0.02)].index.to_list()
manufacturer_list

['bmw',
 'chevrolet',
 'dodge',
 'ford',
 'gmc',
 'honda',
 'hyundai',
 'jeep',
 'mercedes-benz',
 'nissan',
 'ram',
 'subaru',
 'toyota',
 'volkswagen']

In [28]:
manufacturer_filtered_df = df.query('manufacturer in @manufacturer_list')
manufacturer_filtered_df.head()

id  region  price    year manufacturer                     model  \
27  7316814884  auburn  33590  2014.0          gmc  sierra 1500 crew cab slt   
28  7316814758  auburn  22590  2010.0    chevrolet            silverado 1500   
29  7316814989  auburn  39590  2020.0    chevrolet       silverado 1500 crew   
30  7316743432  auburn  30990  2017.0       toyota      tundra double cab sr   
31  7316356412  auburn  15000  2013.0         ford                 f-150 xlt   

    condition    cylinders fuel  odometer title_status transmission  \
27       good  8 cylinders  gas   57923.0        clean        other   
28       good  8 cylinders  gas   71229.0        clean        other   
29       good  8 cylinders  gas   19160.0        clean        other   
30       good  8 cylinders  gas   41124.0        clean        other   
31  excellent  6 cylinders  gas  128000.0        clean    automatic   

                  VIN drive       size    type paint_color state  
27  3GTP1VEC4EG551563   NaN        NaN  pickup       white    al  
28  1GCSCSE06AZ123805   NaN        NaN  pickup        blue    al  
29  3GCPWCED5LG130317   NaN        NaN  pickup         red    al  
30  5TFRM5F17HX120972   NaN        NaN  pickup         red    al  
31                NaN   rwd  full-size   truck       black    al

In [29]:
manufacturer_filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 324129 entries, 27 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            324129 non-null  int64  
 1   region        324129 non-null  object 
 2   price         324129 non-null  int64  
 3   year          324129 non-null  float64
 4   manufacturer  324129 non-null  object 
 5   model         319996 non-null  object 
 6   condition     184695 non-null  object 
 7   cylinders     194211 non-null  object 
 8   fuel          321977 non-null  object 
 9   odometer      320658 non-null  float64
 10  title_status  318001 non-null  object 
 11  transmission  322195 non-null  object 
 12  VIN           200233 non-null  object 
 13  drive         231885 non-null  object 
 14  size          93923 non-null   object 
 15  type          252569 non-null  object 
 16  paint_color   225887 non-null  object 
 17  state         324129 non-null  object 
dtypes: 

##### Model

In [36]:
manufacturer_filtered_df.model.value_counts()[:100]

f-150                        8009
silverado 1500               5140
1500                         4210
camry                        3135
silverado                    3021
                             ... 
focus se                      552
tiguan                        549
wrangler unlimited sahara     534
camry le                      526
f350                          521
Name: model, Length: 100, dtype: int64

Looking at the Top 100 models included as part of this dataset indicates that each model represented includes at least 500 samples for reference. While some of these values are clearly duplicates that will need to be merged as part of data preperation, limiting to these values would help reduce the amount of data used for analysis while still maintaining a significant representation of each model included for model level analysis if desired.

In [41]:
top_100_models_list = manufacturer_filtered_df.model.value_counts()[:100].index.to_list()
top_100_models_list

['f-150',
 'silverado 1500',
 '1500',
 'camry',
 'silverado',
 'accord',
 'wrangler',
 'civic',
 'altima',
 'escape',
 '2500',
 'tacoma',
 'explorer',
 'grand cherokee',
 'corolla',
 'mustang',
 'fusion',
 'equinox',
 'cr-v',
 'focus',
 'malibu',
 'tahoe',
 'charger',
 'corvette',
 'rav4',
 'sonata',
 'impala',
 'sierra 1500',
 'grand caravan',
 'cruze',
 'outback',
 'silverado 2500hd',
 'f-250',
 'elantra',
 'odyssey',
 '3500',
 'edge',
 'prius',
 'tundra',
 'jetta',
 'rogue',
 'forester',
 'sentra',
 'sierra',
 'wrangler unlimited',
 'camaro',
 'pilot',
 'c-class',
 'sienna',
 'f-350',
 'f150',
 'cherokee',
 '4runner',
 'highlander',
 'suburban',
 'expedition',
 'traverse',
 'impreza',
 'challenger',
 'acadia',
 '3 series',
 'sierra 2500hd',
 'e-class',
 'yukon',
 'x5',
 'santa fe',
 'taurus',
 'ranger',
 'terrain',
 'colorado',
 'durango',
 'f250',
 'journey',
 'passat',
 'maxima',
 'pathfinder',
 'liberty',
 'versa',
 'murano',
 '5 series',
 'patriot',
 'fusion se',
 'legacy',
 'co

In [42]:
top_100_models_df = manufacturer_filtered_df.query('model in @top_100_models_list')
top_100_models_df.head()

id  region  price    year manufacturer                     model  \
28  7316814758  auburn  22590  2010.0    chevrolet            silverado 1500   
34  7316285779  auburn  35000  2019.0       toyota                    tacoma   
37  7316130053  auburn   4500  1992.0         jeep                  cherokee   
38  7315816316  auburn  32990  2017.0         jeep  wrangler unlimited sport   
39  7315770394  auburn  24590  2017.0    chevrolet    silverado 1500 regular   

    condition    cylinders fuel  odometer title_status transmission  \
28       good  8 cylinders  gas   71229.0        clean        other   
34  excellent  6 cylinders  gas   43000.0        clean    automatic   
37  excellent  6 cylinders  gas  192000.0        clean    automatic   
38       good  6 cylinders  gas   30041.0        clean        other   
39       good  6 cylinders  gas   40784.0        clean        other   

                  VIN drive size    type paint_color state  
28  1GCSCSE06AZ123805   NaN  NaN  pickup        blue    al  
34                NaN   4wd  NaN   truck        grey    al  
37                NaN   4wd  NaN     NaN         NaN    al  
38  1C4BJWDG5HL705371   4wd  NaN   other      silver    al  
39  1GCNCNEH7HZ118773   NaN  NaN  pickup       white    al

In [43]:
top_100_models_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 141742 entries, 28 to 426856
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            141742 non-null  int64  
 1   region        141742 non-null  object 
 2   price         141742 non-null  int64  
 3   year          141742 non-null  float64
 4   manufacturer  141742 non-null  object 
 5   model         141742 non-null  object 
 6   condition     65407 non-null   object 
 7   cylinders     80161 non-null   object 
 8   fuel          141121 non-null  object 
 9   odometer      139644 non-null  float64
 10  title_status  140586 non-null  object 
 11  transmission  140459 non-null  object 
 12  VIN           84532 non-null   object 
 13  drive         104212 non-null  object 
 14  size          44426 non-null   object 
 15  type          107153 non-null  object 
 16  paint_color   97130 non-null   object 
 17  state         141742 non-null  object 
dtypes: 

##### Condition

In [45]:
top_100_models_df.condition.value_counts()

excellent    35734
good         20550
like new      6288
fair          2301
new            331
salvage        203
Name: condition, dtype: int64

##### Cylinders

In [46]:
top_100_models_df.cylinders.value_counts()

4 cylinders     29255
6 cylinders     27767
8 cylinders     22141
5 cylinders       390
10 cylinders      279
other             246
3 cylinders        79
12 cylinders        4
Name: cylinders, dtype: int64

Cars with a cylinder number outside of 4, 6, 8 represent an unusual car. As such, for the purposes of modeling it is better to filter these cars out as the price to car regression is likely to be significantly different from standard cars.

In [47]:
normal_cylinders_list = ['4 cylinders', '6 cylinders', '8 cylinders']
normal_cylinders_df = top_100_models_df.query('cylinders in @normal_cylinders_list')
normal_cylinders_df.head()

id  region  price    year manufacturer                     model  \
28  7316814758  auburn  22590  2010.0    chevrolet            silverado 1500   
34  7316285779  auburn  35000  2019.0       toyota                    tacoma   
37  7316130053  auburn   4500  1992.0         jeep                  cherokee   
38  7315816316  auburn  32990  2017.0         jeep  wrangler unlimited sport   
39  7315770394  auburn  24590  2017.0    chevrolet    silverado 1500 regular   

    condition    cylinders fuel  odometer title_status transmission  \
28       good  8 cylinders  gas   71229.0        clean        other   
34  excellent  6 cylinders  gas   43000.0        clean    automatic   
37  excellent  6 cylinders  gas  192000.0        clean    automatic   
38       good  6 cylinders  gas   30041.0        clean        other   
39       good  6 cylinders  gas   40784.0        clean        other   

                  VIN drive size    type paint_color state  
28  1GCSCSE06AZ123805   NaN  NaN  pickup        blue    al  
34                NaN   4wd  NaN   truck        grey    al  
37                NaN   4wd  NaN     NaN         NaN    al  
38  1C4BJWDG5HL705371   4wd  NaN   other      silver    al  
39  1GCNCNEH7HZ118773   NaN  NaN  pickup       white    al

In [48]:
normal_cylinders_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 79163 entries, 28 to 426836
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            79163 non-null  int64  
 1   region        79163 non-null  object 
 2   price         79163 non-null  int64  
 3   year          79163 non-null  float64
 4   manufacturer  79163 non-null  object 
 5   model         79163 non-null  object 
 6   condition     51064 non-null  object 
 7   cylinders     79163 non-null  object 
 8   fuel          78976 non-null  object 
 9   odometer      78604 non-null  float64
 10  title_status  78553 non-null  object 
 11  transmission  79073 non-null  object 
 12  VIN           43925 non-null  object 
 13  drive         72098 non-null  object 
 14  size          38234 non-null  object 
 15  type          70180 non-null  object 
 16  paint_color   65028 non-null  object 
 17  state         79163 non-null  object 
dtypes: float64(2), int64(2),

##### Fuel

In [49]:
normal_cylinders_df.fuel.value_counts()

gas         72688
diesel       4484
hybrid        934
other         866
electric        4
Name: fuel, dtype: int64

The majority of cars are of the **"gas"** fuel type. While it could also be argued that **"diesel"** represents a significant amount of sample data, given the current time constraint limits and the fact that analyizing a separate fuel type would jusitfy separate modeling, for the purposes of the initial cycle of analysis data will be limited to only the **"gas"** type.

In [50]:
gas_df = normal_cylinders_df[normal_cylinders_df['fuel'] == 'gas']
gas_df.head()

id  region  price    year manufacturer                     model  \
28  7316814758  auburn  22590  2010.0    chevrolet            silverado 1500   
34  7316285779  auburn  35000  2019.0       toyota                    tacoma   
37  7316130053  auburn   4500  1992.0         jeep                  cherokee   
38  7315816316  auburn  32990  2017.0         jeep  wrangler unlimited sport   
39  7315770394  auburn  24590  2017.0    chevrolet    silverado 1500 regular   

    condition    cylinders fuel  odometer title_status transmission  \
28       good  8 cylinders  gas   71229.0        clean        other   
34  excellent  6 cylinders  gas   43000.0        clean    automatic   
37  excellent  6 cylinders  gas  192000.0        clean    automatic   
38       good  6 cylinders  gas   30041.0        clean        other   
39       good  6 cylinders  gas   40784.0        clean        other   

                  VIN drive size    type paint_color state  
28  1GCSCSE06AZ123805   NaN  NaN  pickup        blue    al  
34                NaN   4wd  NaN   truck        grey    al  
37                NaN   4wd  NaN     NaN         NaN    al  
38  1C4BJWDG5HL705371   4wd  NaN   other      silver    al  
39  1GCNCNEH7HZ118773   NaN  NaN  pickup       white    al

In [51]:
gas_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72688 entries, 28 to 426813
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            72688 non-null  int64  
 1   region        72688 non-null  object 
 2   price         72688 non-null  int64  
 3   year          72688 non-null  float64
 4   manufacturer  72688 non-null  object 
 5   model         72688 non-null  object 
 6   condition     47180 non-null  object 
 7   cylinders     72688 non-null  object 
 8   fuel          72688 non-null  object 
 9   odometer      72148 non-null  float64
 10  title_status  72170 non-null  object 
 11  transmission  72601 non-null  object 
 12  VIN           39622 non-null  object 
 13  drive         65970 non-null  object 
 14  size          35524 non-null  object 
 15  type          64467 non-null  object 
 16  paint_color   60269 non-null  object 
 17  state         72688 non-null  object 
dtypes: float64(2), int64(2),

##### Title Status

In [52]:
gas_df.title_status.value_counts()

clean         69042
rebuilt        1731
salvage         932
lien            376
missing          63
parts only       26
Name: title_status, dtype: int64

While there is merit in exploring cars with different title statuses, this would justify separate modeling as the value assessed to a given car would likely be significantly different. As such, given the current time constraints, this analysis will focus on cars with only **"clean"** status

In [53]:
clean_title_df = gas_df[gas_df['title_status'] == 'clean']
clean_title_df.head()

id  region  price    year manufacturer                     model  \
28  7316814758  auburn  22590  2010.0    chevrolet            silverado 1500   
34  7316285779  auburn  35000  2019.0       toyota                    tacoma   
37  7316130053  auburn   4500  1992.0         jeep                  cherokee   
38  7315816316  auburn  32990  2017.0         jeep  wrangler unlimited sport   
39  7315770394  auburn  24590  2017.0    chevrolet    silverado 1500 regular   

    condition    cylinders fuel  odometer title_status transmission  \
28       good  8 cylinders  gas   71229.0        clean        other   
34  excellent  6 cylinders  gas   43000.0        clean    automatic   
37  excellent  6 cylinders  gas  192000.0        clean    automatic   
38       good  6 cylinders  gas   30041.0        clean        other   
39       good  6 cylinders  gas   40784.0        clean        other   

                  VIN drive size    type paint_color state  
28  1GCSCSE06AZ123805   NaN  NaN  pickup        blue    al  
34                NaN   4wd  NaN   truck        grey    al  
37                NaN   4wd  NaN     NaN         NaN    al  
38  1C4BJWDG5HL705371   4wd  NaN   other      silver    al  
39  1GCNCNEH7HZ118773   NaN  NaN  pickup       white    al

In [54]:
clean_title_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69042 entries, 28 to 426813
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            69042 non-null  int64  
 1   region        69042 non-null  object 
 2   price         69042 non-null  int64  
 3   year          69042 non-null  float64
 4   manufacturer  69042 non-null  object 
 5   model         69042 non-null  object 
 6   condition     44314 non-null  object 
 7   cylinders     69042 non-null  object 
 8   fuel          69042 non-null  object 
 9   odometer      68502 non-null  float64
 10  title_status  69042 non-null  object 
 11  transmission  68955 non-null  object 
 12  VIN           38267 non-null  object 
 13  drive         62922 non-null  object 
 14  size          33726 non-null  object 
 15  type          61349 non-null  object 
 16  paint_color   57301 non-null  object 
 17  state         69042 non-null  object 
dtypes: float64(2), int64(2),

##### Transimission

In [55]:
clean_title_df.transmission.value_counts()

automatic    62459
manual        3994
other         2502
Name: transmission, dtype: int64

Following the same line of reasoning as previous features, analysis will focus on **"automatic"** cars to reduce the need for separate modeling.

In [56]:
automatic_transmission_df = clean_title_df[clean_title_df['transmission'] == 'automatic']
automatic_transmission_df.head()

id      region  price    year manufacturer     model  condition  \
34   7316285779      auburn  35000  2019.0       toyota    tacoma  excellent   
37   7316130053      auburn   4500  1992.0         jeep  cherokee  excellent   
59   7313406529      auburn  14000  2012.0        honda   odyssey  excellent   
95   7309798041      auburn   2100  2006.0       subaru   impreza       fair   
170  7316874816  birmingham  13950  2011.0       toyota    tacoma       good   

       cylinders fuel  odometer title_status transmission  VIN drive  \
34   6 cylinders  gas   43000.0        clean    automatic  NaN   4wd   
37   6 cylinders  gas  192000.0        clean    automatic  NaN   4wd   
59   6 cylinders  gas   95000.0        clean    automatic  NaN   fwd   
95   4 cylinders  gas   97000.0        clean    automatic  NaN   NaN   
170  4 cylinders  gas  151060.0        clean    automatic  NaN   rwd   

          size       type paint_color state  
34         NaN      truck        grey    al  
37         NaN        NaN         NaN    al  
59   full-size   mini-van      silver    al  
95         NaN  hatchback         NaN    al  
170        NaN      truck      silver    al

In [57]:
automatic_transmission_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 62459 entries, 34 to 426813
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            62459 non-null  int64  
 1   region        62459 non-null  object 
 2   price         62459 non-null  int64  
 3   year          62459 non-null  float64
 4   manufacturer  62459 non-null  object 
 5   model         62459 non-null  object 
 6   condition     39649 non-null  object 
 7   cylinders     62459 non-null  object 
 8   fuel          62459 non-null  object 
 9   odometer      61947 non-null  float64
 10  title_status  62459 non-null  object 
 11  transmission  62459 non-null  object 
 12  VIN           34071 non-null  object 
 13  drive         57362 non-null  object 
 14  size          31595 non-null  object 
 15  type          55453 non-null  object 
 16  paint_color   51699 non-null  object 
 17  state         62459 non-null  object 
dtypes: float64(2), int64(2),

##### Drive

In [58]:
automatic_transmission_df.drive.value_counts()

4wd    27277
fwd    20734
rwd     9351
Name: drive, dtype: int64

The relationship of drive to price is unknown at this point in time. As such, the feature will be included as part of model training. As no best guess assumption can be made for samples with a missing drive field, the best course of action is to drop these samples from analysis.

##### Size

In [60]:
automatic_transmission_df['size'].value_counts()

full-size      16645
mid-size       11032
compact         3514
sub-compact      404
Name: size, dtype: int64

While there would be value in using size as part of further analysis, given that samples with a missing size value represent almost half of the remaining samples and no best guess assumption can be made about these missing values, it is better to drop this field for initial model training.

Future analysis cycles could benefit from examining cars with size values explicitly registered.

##### Type

In [63]:
automatic_transmission_df.type.value_counts()

SUV            17837
sedan          15861
truck           9871
pickup          4923
coupe           1975
wagon           1357
mini-van        1063
van              702
hatchback        587
other            571
convertible      557
offroad          147
bus                2
Name: type, dtype: int64

In an effort to reduce noise, cars who's type value count are less than 500 will be removed.

In [71]:
car_type_list = automatic_transmission_df.groupby('type')['type'].filter(lambda col: len(col) > 500).unique()
car_type_list

array(['truck', 'mini-van', 'hatchback', 'coupe', 'sedan', 'SUV',
       'pickup', 'other', 'wagon', 'van', 'convertible'], dtype=object)

In [72]:
car_type_df = automatic_transmission_df.query('type in @car_type_list')
car_type_df.head()

id      region  price    year manufacturer    model  condition  \
34   7316285779      auburn  35000  2019.0       toyota   tacoma  excellent   
59   7313406529      auburn  14000  2012.0        honda  odyssey  excellent   
95   7309798041      auburn   2100  2006.0       subaru  impreza       fair   
170  7316874816  birmingham  13950  2011.0       toyota   tacoma       good   
187  7316846605  birmingham   4500  2008.0         ford  mustang        NaN   

       cylinders fuel  odometer title_status transmission                VIN  \
34   6 cylinders  gas   43000.0        clean    automatic                NaN   
59   6 cylinders  gas   95000.0        clean    automatic                NaN   
95   4 cylinders  gas   97000.0        clean    automatic                NaN   
170  4 cylinders  gas  151060.0        clean    automatic                NaN   
187  6 cylinders  gas  187016.0        clean    automatic  1ZVHT80N885207297   

    drive       size       type paint_color state  
34    4wd        NaN      truck        grey    al  
59    fwd  full-size   mini-van      silver    al  
95    NaN        NaN  hatchback         NaN    al  
170   rwd        NaN      truck      silver    al  
187   rwd        NaN      coupe         red    al

In [73]:
car_type_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55304 entries, 34 to 426784
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            55304 non-null  int64  
 1   region        55304 non-null  object 
 2   price         55304 non-null  int64  
 3   year          55304 non-null  float64
 4   manufacturer  55304 non-null  object 
 5   model         55304 non-null  object 
 6   condition     34147 non-null  object 
 7   cylinders     55304 non-null  object 
 8   fuel          55304 non-null  object 
 9   odometer      54868 non-null  float64
 10  title_status  55304 non-null  object 
 11  transmission  55304 non-null  object 
 12  VIN           32038 non-null  object 
 13  drive         52090 non-null  object 
 14  size          30772 non-null  object 
 15  type          55304 non-null  object 
 16  paint_color   47765 non-null  object 
 17  state         55304 non-null  object 
dtypes: float64(2), int64(2),

##### Paint Color

In [74]:
car_type_df.paint_color.value_counts()

white     10567
black      9005
silver     7313
grey       6267
blue       4954
red        4570
custom     2202
brown      1161
green      1133
yellow      276
orange      199
purple      118
Name: paint_color, dtype: int64

##### State

In [75]:
car_type_df.state.value_counts()

ca    5286
fl    3577
tx    3214
or    2482
nc    2422
ny    2382
oh    2307
mi    1723
va    1699
wi    1627
wa    1624
tn    1601
ia    1569
co    1564
pa    1471
nj    1431
il    1282
id    1269
mn    1202
mt    1140
ma    1028
az     955
ok     897
ga     886
ks     789
md     746
sc     737
ak     663
mo     630
al     623
ct     613
ky     588
in     586
nv     575
vt     514
ar     396
hi     339
nh     331
dc     320
nm     316
la     312
me     308
ne     249
ri     242
sd     211
ms     137
ut     125
de     114
wv      99
nd      54
wy      49
Name: state, dtype: int64

While the state for a given car is likely to have an impact on the price, this is not a feature that the client would be able to have any control over. As such, for the purpose of reducing data complexity, this feature can be dropped.

#### Summary of Exploration
Given time constraints for the current cycle of data analysis, data will be filtered on a number of features:
- Manufacturer: Manufacturers with over 2% representation in dataset
- Model: Top 100 models
- Cylinders: Cars with 4, 6, or 8 cylinders
- Fuel: Cars with "gas" type
- Title: Cars with "clean" status
- Transmission: Automatic
- Type: Cars with type count of over 500

##### Fields to Drop
Based on initial assessment there are a number of fields that have been deemed unneccessary for the initial cycle of analysis and as such will be dropped:
- id
- region
- VIN
- size
- state

##### Fields with Missing Value Handling
After filtering based on the above, there are still a significant number of samples that contain fields with missing values. This includes the following fields:
- condition
- odometer
- drive
- paint_color

The most significant of these being **"condition"**, with a non-null value of 34,147. Given that this is likely a key feature in determining the price of a car and that no best guess assumption can be made for the non-null values, the course of action in this case is for rows with missing values to be dropped.

With the above data cleaning, the final data samples will end up being around 34,147/426,880, or approximately 8% of the initial dataset. While this may seem like a small value, the absolute value is still significantly large such that we should be able to interpret the results of any analysis performed as a reliable model for car price. 

Further, it should be noted that this sample set represents the most common types of cars represented. As such, this should provide us with a solid foundation for determining an appropriate strategy for increasing revenue on the majority of cars sold by the client.

## Data Preparation

After our initial exploration and fine tuning of the business understanding, it is time to construct our final dataset prior to modeling.  Here, we want to make sure to handle any integrity issues and cleaning, the engineering of new features, any transformations that we believe should happen (scaling, logarithms, normalization, etc.), and general preparation for modeling with `sklearn`. 

### Drop unnecessary features, samples with missing values

In [76]:
dropped_features_df = car_type_df.drop(['id', 'region', 'VIN', 'size', 'state'], axis = 1)
dropped_features_df.head()

price    year manufacturer    model  condition    cylinders fuel  \
34   35000  2019.0       toyota   tacoma  excellent  6 cylinders  gas   
59   14000  2012.0        honda  odyssey  excellent  6 cylinders  gas   
95    2100  2006.0       subaru  impreza       fair  4 cylinders  gas   
170  13950  2011.0       toyota   tacoma       good  4 cylinders  gas   
187   4500  2008.0         ford  mustang        NaN  6 cylinders  gas   

     odometer title_status transmission drive       type paint_color  
34    43000.0        clean    automatic   4wd      truck        grey  
59    95000.0        clean    automatic   fwd   mini-van      silver  
95    97000.0        clean    automatic   NaN  hatchback         NaN  
170  151060.0        clean    automatic   rwd      truck      silver  
187  187016.0        clean    automatic   rwd      coupe         red

In [77]:
no_missing_values_df = dropped_features_df.dropna()
no_missing_values_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28900 entries, 34 to 426784
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         28900 non-null  int64  
 1   year          28900 non-null  float64
 2   manufacturer  28900 non-null  object 
 3   model         28900 non-null  object 
 4   condition     28900 non-null  object 
 5   cylinders     28900 non-null  object 
 6   fuel          28900 non-null  object 
 7   odometer      28900 non-null  float64
 8   title_status  28900 non-null  object 
 9   transmission  28900 non-null  object 
 10  drive         28900 non-null  object 
 11  type          28900 non-null  object 
 12  paint_color   28900 non-null  object 
dtypes: float64(2), int64(1), object(10)
memory usage: 3.1+ MB


### Combine common "model" values

In [104]:
# Combine all F150 Cars
no_missing_values_df.loc[no_missing_values_df['model'] == 'f150', 'model'] = 'f-150'

# Combine all 1500 Cars
no_missing_values_df.loc[no_missing_values_df['model'] == 'silverado 1500', 'model'] = '1500'
no_missing_values_df.loc[no_missing_values_df['model'] == 'sierra 1500', 'model'] = '1500'
no_missing_values_df.loc[no_missing_values_df['model'] == 'sierra 1500', 'model'] = '1500'
no_missing_values_df.loc[no_missing_values_df['model'] == '1500 crew cab big horn', 'model'] = '1500'

# Combine all 2500 Cars
no_missing_values_df.loc[no_missing_values_df['model'] == 'silverado 2500hd', 'model'] = '2500'
no_missing_values_df.loc[no_missing_values_df['model'] == 'sierra 2500hd', 'model'] = '2500'

# Combine all F250 Cars
no_missing_values_df.loc[no_missing_values_df['model'] == 'f250', 'model'] = 'f-250'
no_missing_values_df.loc[no_missing_values_df['model'] == 'f250 super duty', 'model'] = 'f-250'
no_missing_values_df.loc[no_missing_values_df['model'] == 'f-250 super duty', 'model'] = 'f-250'

# Combine all F350 Cars
no_missing_values_df.loc[no_missing_values_df['model'] == 'f350', 'model'] = 'f-350'
no_missing_values_df.loc[no_missing_values_df['model'] == 'super duty f-350 srw', 'model'] = 'f-350'

### Remove cars with zero price value
During preprocessing setup it was discovered that samples exist where the price value is set to 0. This would appear to be faulty data, or suggest that these vehicles were given away for free. Either way, given that we want to examine cars that were sold we will remove these samples from the dataset.

Follow up investigation into why such samples exist is recommended.

In [132]:
no_free_cars_df = no_missing_values_df[no_missing_values_df['price'] != 0]
no_free_cars_df[no_free_cars_df['price'] == 0]

Empty DataFrame
Columns: [price, year, manufacturer, model, condition, cylinders, fuel, odometer, title_status, transmission, drive, type, paint_color]
Index: []

### Preprocess Encoding and Transformation
#### Encoding
Given that categorical features are non-ordinal, One Hot Encoding will be used on all categorical features. Given the high number of categories in the **"model"** feature, there is a concern on the impact this feature will have on dimensional complexity.

The number of categories for this feature will be left unchanged for now but if a significant impact on processing time is observed the number of models included will be reduced.

#### Transformation
For the numerical features **"year"** and **"odometer"**, two transofrmations will be performed:
1. StandardScalar - This is to ensure that one feature doesn't heavily outweigh the other due to a larger variance in values.
2. PolynomialFeatures - To explore if higher order variations of these values act as a more effective predictor of price.

Hyperparameters for PolynomialFeatures and other model relevant handlers will be optimized as part of a Grid Search during the Modeling Phase. As such, only default values will be used at this stage, defining and confirming behavior for the base preprocessing pipeline.

In [133]:
# Use pipeline to perform multiple transformations on specific columns (see transformer below)
num_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('poly', PolynomialFeatures(include_bias = False))
])

transformer = make_column_transformer(
    (OneHotEncoder(), [
        'manufacturer', 
        'model', 
        'condition', 
        'cylinders', 
        'fuel', 
        'title_status', 
        'transmission', 
        'drive', 
        'type', 
        'paint_color'
    ]),
    (num_pipeline, ['year', 'odometer']),
    verbose_feature_names_out = False,
    remainder = 'passthrough',
    sparse_threshold = 0 # Due to current lack of knowledge of working with sparse matrices, dense matrix return will be enforced
)

# Test to make sure it functions as expected
transformed = transformer.fit_transform(no_free_cars_df)
transformed_df = pd.DataFrame(transformed, columns=transformer.get_feature_names_out())
transformed_df.head()

manufacturer_bmw  manufacturer_chevrolet  manufacturer_dodge  \
0               0.0                     0.0                 0.0   
1               0.0                     0.0                 0.0   
2               0.0                     0.0                 0.0   
3               0.0                     0.0                 0.0   
4               0.0                     1.0                 0.0   

   manufacturer_ford  manufacturer_gmc  manufacturer_honda  \
0                0.0               0.0                 0.0   
1                0.0               0.0                 1.0   
2                0.0               0.0                 0.0   
3                0.0               0.0                 1.0   
4                0.0               0.0                 0.0   

   manufacturer_hyundai  manufacturer_jeep  manufacturer_mercedes-benz  \
0                   0.0                0.0                         0.0   
1                   0.0                0.0                         0.0   
2                   0.0                0.0                         0.0   
3                   0.0                0.0                         0.0   
4                   0.0                0.0                         0.0   

   manufacturer_nissan  ...  paint_color_red  paint_color_silver  \
0                  0.0  ...              0.0                 0.0   
1                  0.0  ...              0.0                 1.0   
2                  0.0  ...              0.0                 1.0   
3                  0.0  ...              0.0                 1.0   
4                  0.0  ...              1.0                 0.0   

   paint_color_white  paint_color_yellow      year  odometer    year^2  \
0                0.0                 0.0  1.306524 -0.637657  1.707006   
1                0.0                 0.0  0.302490 -0.223964  0.091500   
2                0.0                 0.0  0.159057  0.222029  0.025299   
3                0.0                 0.0 -0.271243  0.647663  0.073573   
4                0.0                 0.0  1.019658 -0.247831  1.039702   

   year odometer  odometer^2    price  
0      -0.833115    0.406607  35000.0  
1      -0.067747    0.050160  14000.0  
2       0.035315    0.049297  13950.0  
3      -0.175674    0.419468   2500.0  
4      -0.252703    0.061420  45000.0  

[5 rows x 146 columns]

## Modeling

With your (almost?) final dataset in hand, it is now time to build some models.  Here, you should build a number of different regression models with the price as the target.  In building your models, you should explore different parameters and be sure to cross-validate your findings.

### Evaluation

With some modeling accomplished, we aim to reflect on what we identify as a high quality model and what we are able to learn from this.  We should review our business objective and explore how well we can provide meaningful insight on drivers of used car prices.  Your goal now is to distill your findings and determine whether the earlier phases need revisitation and adjustment or if you have information of value to bring back to your client.

## Deployment

Now that we've settled on our models and findings, it is time to deliver the information to the client.  You should organize your work as a basic report that details your primary findings.  Keep in mind that your audience is a group of used car dealers interested in fine tuning their inventory.